### ISSUE TITLE
goroutines and channels

### ISSUE THREAD

**Issue by glycerine** [go-interpreter/gi#10]
Goroutines and channels are nice-to-haves, long term aims. I'm not worrying about them in the initial implementation of `gi`. There is enough to do just getting straight-line code to work. And its worth reviewing how Elliott dealt with them in the haxe runtime when it comes to it.

Nonetheless, there is a language, shine[1], that compiles to Lua that has implemented some CSP constructs, namely fibers, so shine could be a useful resource when it comes to implementing these; and of course Lua does have built in actual coroutines.

Shine also implements classes. These may or may not be helpful when it comes to doing interfaces.

[1] https://github.com/richardhundt/shine#concurrency

**Comment by glycerine** [comment]
Actually this would be an excellent area for somebody beside myself to take on, since there is very little overlap with my current focus.

**Comment by elliott5** [comment]
To implement goroutines in a single-thread environment will require [coroutines](https://en.wikipedia.org/wiki/Coroutine).

Any implementation has to keep a "stack" for each goroutine, where the flow of control passes between the various "stacks" depending on which goroutine is active. The "stack" will contain stack-frames for each function/method on that stack, each stack-frame represented by a program-counter of some kind within the function/method and the parameters/local variables/return values used by the function/method. 

The key issue is at which points in the code to pass control to other goroutines (usually by returning from an actual subroutine to the main goroutine/coroutine  scheduler).

The operation of the stack is complicated by Go's defer/recover functionality.

I hope it is clear from my brief outline above that this coroutine approach affects all flows of control and stack-based variables, I doubt it can be done as an add-on to the main project.

That said, maybe Shine Lua dialect could do most of the heavy lifting by using the Fibre class, provided the defer/recover logic were handled elsewhere.

The place to start would be by considering how to model channels in Shine Lua...

**Comment by glycerine** [comment]
A killer feature would be to send on channels between Go libs and Lua (interpreted code). It would need CGO, but we're already doing CGO for LuaJIT.  Bootstrapping this would need some kind of registry for rendezvous, but it all seems quite doable. And awesome.

**Comment by elliott5** [comment]
Yes, the lua-channels library may be a better alternative to Shine. As I have never programmed in Lua I am unable to judge between them.

Sending/receiving on native Go channels to/from Lua, or rather code running inside the Lua JIT called by CGo will be challenging (given all of the CGo call is inside a single goroutine), but perhaps not impossible eventually.

Sadly Jason, this is not a space in which I feel qualified to help (beyond cheering-you-on and offering dubious advice), both because I do not know Lua, and because I've no need for a Go interpreter in my yet-to-be-a-startup project (so cannot justify the time away from what I should be doing).

**Comment by glycerine** [comment]
Ooooohh hold up. Steve Donovan's LuaR project claims to be able to both import Go libs into Lua, and to be able to send on channels.

https://github.com/stevedonovan/luar

He claims to have Lua object <-> Go struct transforms in place both directions. So if `gi` generates Lua objects directly that match their intended native Go struct, then we would have automatic access to all of the Go native standard library and scheduler.  Which would be a big win.

**Comment by elliott5** [comment]
Maybe LuaR points to a different approach for writing gi, which I would
think would be likely to me more acceptable to the Go community than
translation go->Lua.

Only use the Lua virtual machine to hold the data between REPL
interactions, not the code.

In this model Lua becomes a persistent data store which native Go accesses.

When a new Go fragment were typed into the REPL for execution, gi would:
1) Create a Go source file with a unique package name ("repl-line-42")
containing code reflecting the environment defined by previous REPL
statements.
2) Add the entered code fragment to the end of that file, compile it using
the normal go compiler as a plugin <https://golang.org/pkg/plugin/>.
3) Load the generated plugin code and run it through LuaR.

This has the advantages that all but the top-level glue code runs as native
Go, which is likely to be more acceptable to the community.

It has the disadvantage that plugin mode is currently only available in
Linux (so the REPL would need to run as a Docker container in OSX/Windows).

The translation go->lua that you have been working on could maybe still be
used for simple statements without control flow at the REPL command prompt.

I've been thinking about this approach to writing a Go REPL for a while, it
seems LuaR would make it possible.

**Comment by glycerine** [comment]
I think there are a couple of extant projects that kind of take that approach, no? neugram (with plugins), gore (pre plugins), others.

I'm not sure it will work to use the current Go native compiler for every run.  Consider: I don't want to re-run all the prior code before running any new code. The older code may have computed for hours, and my new line of code may be a simple print statement. Without the earlier code, the new code won't type check and compile by the Go compiler (unmodified that is). 

The type checker is the real nub of the problem, actually. 

For example, `gi` allows expressions at the top level (package scope, effectively) so that one can print out the values just defined. But `gc` would fail those as illegal Go.  One can't say to `gc`
~~~
package main

var a = 2

println(a)
~~~
because `gc` will object that isn't up to snuff. But `gi` happily prints 2 for us.

Or perhaps you were thinking of modifying and re-building the `gc` compiler itself? That could indeed be quite an intriguing route... have to think on that more.

LuaJIT grows on one with more experience, and Lua has the ability to emulate quite sophisticated object systems and method overloading (much beyond what Go offers or requires). I think that having even a standalone full Go -> full Lua will be incredibly valuable. It matches what GopherJS does for full Go -> full Javascript. But I do think the opportunities for integration via LuaR with the standard library are too great to pass up as well, and so want to design with the eventuality of that integration in mind.

**Comment by elliott5** [comment]
I was thinking in terms of only recompiling the latest go fragment as a plug-in, with that fragment rewritten to access all of the existing data (which may have been created over days as you say). So not recompiling the whole program each time.

Each of the top level REPL go fragments would be run via Lua, so their output would be accessible by future go fragments using Lua as the glue. 

Think of it as a LuaR program where every Statement is to run a function in a native go plugin in the way you have described.

In terms of what a REPL is in Go terms, I think it can only be a main() function where the code is compiled over time. Plus the ability to import packages retrospectively of course.

You’re closer to the problem than I am, so better able to judge if LuaR really is a game-changer in this regard.
